pipeline/run_pipeline.py

```diff
@@ -39,186 +93,533 @@ def load_and_sample_datasets(cfg):
+    plt.savefig(plot_file_path)
+    plt.close()
+    print(f"Plot saved to {plot_file_path}")
+
+
+def computing_dot_product(
+    data_list,
+    refusal_direction,
+    average_dot_products,
+    resultant_dot_products,
+    activations_dir,
+    layers,
+    harmful_train,
+):
+    data_list = [normalize_vector(data) for data in data_list]
+
+    """Calculate and save dot products, then generate and save a DataFrame with results."""
+    for i in range(layers):
+        # Calculate average dot products
+        dot_products = sum(
+            np.dot(data_list[j][i].cpu().numpy(), refusal_direction.cpu().numpy())
+            for j in range(len(harmful_train))
+        ) / len(harmful_train)
+        average_dot_products.append(dot_products)
+
+        # Calculate resultant dot products
+        resultant_vector = sum(data_list[j][i] for j in range(len(harmful_train)))
+        resultant_vector = normalize_vector(resultant_vector)
+        dot_product = np.dot(
+            resultant_vector.cpu().numpy(), refusal_direction.cpu().numpy()
+        )
+        resultant_dot_products.append(dot_product)
+
+    print("Average Dot Products:", average_dot_products)
+    print("Resultant Dot Products:", resultant_dot_products)
+
+    # Create and display a DataFrame with results
+    data = {
+        "Layer": list(range(layers)),
+        "Sum along cos(A_i, R)": average_dot_products,
+        "Cos(resultant(A_i), R)": resultant_dot_products,
+    }
+    df = pd.DataFrame(data)
+
+    # Save average_dot_products
+    average_activation_file_path = os.path.join(
+        activations_dir, "average_dot_products.npy"
+    )
+    np.save(average_activation_file_path, average_dot_products)
+
+    # Save resultant_dot_products
+    resultant_activation_file_path = os.path.join(
+        activations_dir, "resultant_dot_products.npy"
+    )
+    np.save(resultant_activation_file_path, resultant_dot_products)
+
+    # Save DataFrame as CSV
+    df_file_path = os.path.join(activations_dir, "cosine_similarity_results.csv")
+    df.to_csv(df_file_path, index=False)
+
+    print("Data saved successfully.")
+    print(df)
+
+    return average_dot_products, resultant_dot_products
+
+
+def generate_and_save_activations(cfg, model_base, harmful_train, refusal_direction):
+    """Generate and save candidate directions."""
+    activations_dir = os.path.join(cfg.artifact_path(), "generate_activations")
+
+    # Create the directory if it doesn't exist
+    if not os.path.exists(activations_dir):
+        os.makedirs(activations_dir)
+
+    # for storing cos with refusal_direction
+    data_list = []
+
+    for i in range(len(harmful_train)):
+        mean_activations = generate_activations_for_harmful(
+            model_base,
+            harmful_train[i],
+        )
+        file_path = os.path.join(activations_dir, f"mean_activations_for_{i}_prompt.pt")
+
+        # mean activations has a size of 5*18*d_model -> it takes activations of last 5 positions of prompt
```

Comment:
this comment should be an assert statement. Actually check within the code whether `mean_activations` has the shape that you expect.